### PR TITLE
Improve Styx data plane.

### PR DESCRIPTION
**Pull request by mikkokar**
# Summary

This PR fixes three little problems that involve responses that arrive before the corresponding request is fully sent.

## Fix 1:
HttpRequestOperation has two callback handlers on the same channel future that both trigger an observable error. Omitting a second `onError` violates Rx.Java protocol. Also it is pointless because Rx suppresses the 2nd `onError` event. 

Impact: Confusing log messages.

Fix: collapse the two callbacks into one. 

## Fix 2:
"Connection: close" header was not included in some of the error responses. 

Impact: The remote peer may start sending the next request on a channel that was about to be closed, resulting in a wasted request.

Fix: Add missing "Connection: close" header to the affected error responses.

## Fix 3:

`HttpRequestOperation.RequestBodyChunkSubscriber` was setting the request body `completed` flag too soon. Therefore `isOngoingRequest` test in `execute` was prematurely returning false. This could allow the connection to be handed over to the next request before the previous is fully sent, resulting in Netty HTTP codec getting out-of-sync.

Impact: A failure to proxy a request to origins. Happens very rarely under heavy load. This issue may show up in load tests.

Fix: Wait until Netty has acknowledged the request EMPTY_LAST_CONTENT before setting completed flag to true.

### CODE REVIEW

File: components/client/src/main/java/com/hotels/styx/client/netty/connectionpool/HttpRequestOperation.java [hunk 1]
```diff
@@ -321,11 +314,20 @@ public void onError(Throwable e) {
         public void onNext(ByteBuf chunk) {
             HttpObject msg = new DefaultHttpContent(chunk);
             channel.writeAndFlush(msg)
-                    .addListener(readNext)
-                    .addListener(readOrClose);
+                    .addListener((ChannelFuture future) -> {
+                        request(1);
```
Comment (dvlato) [inline components/client/src/main/java/com/hotels/styx/client/netty/connectionpool/HttpRequestOperation.java]: Should we encapsulate this in a function?

Comment (mikkokar) [reply components/client/src/main/java/com/hotels/styx/client/netty/connectionpool/HttpRequestOperation.java]: I wouldn't bother. This class is going to get removed quite soon. 🤞

File: components/client/src/main/java/com/hotels/styx/client/netty/connectionpool/HttpRequestOperation.java [hunk 2]
```diff
@@ -156,6 +156,14 @@ private static boolean requestIsOngoing(RequestBodyChunkSubscriber bodyChunkSubs
                         Requests.doFinally(response, cause -> {
                             if (nettyConnection.isConnected()) {
                                 removeProxyBridgeHandlers(nettyConnection);
+
+                                // MK: Request is no longer ongoing. The request body onComplete event has arrived,
```
[the remark this answers is not in the document]
Comment (mikkokar) [reply components/client/src/main/java/com/hotels/styx/client/netty/connectionpool/HttpRequestOperation.java]: I have removed these comments.

File: components/client/src/main/java/com/hotels/styx/client/netty/connectionpool/HttpRequestOperation.java [hunk 3]
```diff
@@ -308,8 +298,11 @@ public void onStart() {
 
         @Override
         public void onCompleted() {
-            completed = true;
-            channel.writeAndFlush(EMPTY_LAST_CONTENT);
+            // MK: This comes from a server thread, and Netty therefore queues it to the Client thread executor.
+            //     But "completed" flag is set to true regardless. Really it should be considered "completed"
```
[the remark this answers is not in the document]
Comment (mikkokar) [reply components/client/src/main/java/com/hotels/styx/client/netty/connectionpool/HttpRequestOperation.java]: 👍

**Pull request merged**